dcicutils/creds_utils.py

```diff
@@ -0,0 +1,272 @@
+# import contextlib
+# import functools
+import contextlib
+import io
+import json
+import os
+
+# from dcicutils.misc_utils import local_attrs
+from dcicutils.exceptions import AppEnvKeyMissing, AppServerKeyMissing
+from dcicutils.misc_utils import remove_suffix
+
+
+# LOCAL_SERVER = "http://localhost:8000"
+# LOCAL_PSEUDOENV = 'fourfront-cgaplocal'
+#
+# PRODUCTION_SERVER = 'https://cgap.hms.harvard.edu'
+# PRODUCTION_ENV = 'fourfront-cgap'
+#
+# keys_file_var='CGAP_KEYS_FILE'
+
+
+_KEY_MANAGERS = {}
+
+
+class KeyManager:
+
+    APP_NAME = None
+    APP_TOKEN = None  # Set this to a string to override APP_NAME.lower() as the app token
+
+    KEYS_FILE_VAR = None
+
+    _REGISTERED = False
+
+    # Instance Methods
+
+    def __init__(self, keys_file=None):
+        if not self._REGISTERED:
+            raise RuntimeError("Only registered KeyManagers can be instantiated.")
+        self.keys_file = keys_file or os.environ.get(self.KEYS_FILE_VAR) or self.KEYS_FILE
+        if not self.keys_file:
+            raise ValueError(f"No KEYS_FILE attribute in {self}, and no {self.KEYS_FILE_VAR} environment variable.")
+
+    KEYS_FILE = None
+    # @property
+    # def KEYS_FILE(self):  # noQA
+    #     # By default this will be computed dynamically, but a KeyManager class can declare a static value by doing
+    #     # something like this:
+    #     #
+    #     #     class MyKeyManager(KeyManager):
+    #     #         KEYS_FILE = 'some file'
+    #     #
+    #     return self._default_keys_file()
+
+    @classmethod
+    @contextlib.contextmanager
+    def default_keys_file_for_testing(cls, filename):
+        """
+        Sets the default keys file for in cls.KEYS_FILE to the indicated filename.
+        Ordinarily, in a non-testing environment, one would set an environment variable to do this,
+        but in the testing environment that has been set already and to a value that has nothing to
+        do with testing. So this bypasses the normal environment variable setup and sets it directly
+        in the class and only for the duration of an evaluation context.
+        """
+
+        old_filename = cls.KEYS_FILE
+        try:
+            cls.KEYS_FILE = filename or cls._default_keys_file(for_testing=True)
+            yield
+        finally:
+            cls.KEYS_FILE = old_filename
+
+    @classmethod
+    def register(cls, *, name):
+        def _register_class(key_manager_class):
```

Comment:
I'd add docstrings especially to functions like these where the syntax/intended structure/use are not immediately obvious (to all) by reading it